### ISSUE TITLE
TipsViewModel lags reality

### ISSUE THREAD

**Issue by gjeee** [iotaledger/iri#761]
### Bug description
When a new tip comes in, it is added by default to the non-solid tip queue (see TransactionValidator.updateStatus()). The TipsManager.scanTipsForSolidity() thread runs every 750 ms. This will pick 1 non-solid tip randomly and checks if it has become solid in the meantime. If this is the case, the tip is put into the solid queue (and removed from the non-solid queue). Now let's say the network runs at 10 tx/s and suppose in this last second these 10 tips all became solid. However the solidity thread in this same timeframe only updates 1 of these. So the TipsViewModel has difficulty keeping up with the current state of things. This has repercussions for the rest of the network as well. Since now and then a neighbor will ask your node for a random tip. Your node will reply then based on its solid tip-view model. But that model misses a lot of recent transactions because they have not been scanned for solidity yet.

### Potential solution

1. Check and update whether a tip is solid immediately after arrival
2. Make the solidity thread do more scans (possibly depending on the size of the non-solid tip queue) and/or more frequent scans

### Am I planning to do it myself with a PR?
Yes / No.

**Comment by footloosejava** [comment]
I am wondering why it runs every 750 ms as well. 

However, let me share this .... the thread does seem very busy when it runs. So I am not sure how many more tips you can scan in that period.

I suspect your CPU usage creep was from Milestone scanning and Tips scanning. As the caches fill, the scanning for both these operations almost becomes full-time.

**Comment by gjeee** [comment]
does it take a lot of resources? that for sure is interesting if you combine it with #756. option 1 seems wise since (i think) 99% of the arriving tx can be "solidity checked" in a quick way....see TransactionValidator.quickSetSolid() routine

**Comment by footloosejava** [comment]
So , do this, change it to check twice or three times, and then sleep 750 ms. See then as an experiment,

**Comment by gjeee** [comment]
That "deepscan" might also use that time-consuming transactionviewmodel call (see #757) which can explain the cost of this call.

**Comment by footloosejava** [comment]
If you can, change it to this, run it for a while and them share how much time is between each number printed on average:

`
    private void scanTipsForSolidity() throws Exception {
        for (int n = 0; n < 10 && tipsViewModel.nonSolidSize() != 0; n++) {
            log.info("checking {}", n);
            Hash hash = tipsViewModel.getRandomNonSolidTipHash();
            if (hash != null) {
                ApproveeViewModel approvers = ApproveeViewModel.load(tangle, hash);
                if (approvers.size() != 0) {
                    tipsViewModel.removeTipHash(hash);
                } else if (transactionValidator.checkSolidity(hash, false)) {
                    tipsViewModel.setSolid(hash);
                }
            }
        }
    }
`

**Comment by footloosejava** [comment]
@gjeee That will show you how many tips can be processed per second. Which, sorry to say is not many.
And yes, tips can come in at a much faster pace than they can be processed. ...

**Comment by gjeee** [comment]
@footloosejava deep-scan is slow indeed. As an alternative one can do a  quick check by leveraging its recursive nature. Just check whether its 2 children are solid. But the deep-scan might be improved performance-wise who knows.

**Comment by gjeee** [comment]
ok i did it slightly different. but the result so far: 280120 ns = 0.28012 milliseconds per tx

```
    private void scanTipsForSolidity() throws Exception {
        int aantalLoops=0;
        long time = System.nanoTime();
	for (int n = 0; n < 10 && tipsViewModel.nonSolidSize() != 0; n++) {
            aantalLoops++;
            Hash hash = tipsViewModel.getRandomNonSolidTipHash();
            if (hash != null) {
                ApproveeViewModel approvers = ApproveeViewModel.load(tangle, hash);
                if (approvers.size() != 0) {
                    tipsViewModel.removeTipHash(hash);
                } else if (transactionValidator.checkSolidity(hash, false)) {
                    tipsViewModel.setSolid(hash);
                }
            }
        }
        if(aantalLoops>0){
	    log.info((System.nanoTime() - time)/aantalLoops + "ns per tx");
	}
    }
```

**Comment by footloosejava** [comment]
After 10 minutes, mine is at 430 ms per scan.

**Comment by footloosejava** [comment]
```
public void init() {
        solidityRescanHandle = new Thread(() -> {

            long totalTime = 0;
            long count = 0;
            while (!shuttingDown) {
                try {
                    count++;
                    long time = System.currentTimeMillis();
                    scanTipsForSolidity();
                    totalTime += System.currentTimeMillis() - time;
                    if (count == 100) {
                        log.info("Last {} scans - avg time= {} ms.", count, (int) (totalTime / count));
                        count = 0;
                        totalTime = 0;
                    }
                } catch (Exception e) {
                    log.error("Error during solidity scan : {}", e);
                }
                try {
                    Thread.sleep(RESCAN_TX_TO_REQUEST_INTERVAL);
                } catch (InterruptedException e) {
                    log.error("Solidity rescan interrupted.");
                }
            }
        }, "Tip Solidity Rescan");
        solidityRescanHandle.start();
    }

    private void scanTipsForSolidity() throws Exception {
        Hash hash = tipsViewModel.getRandomNonSolidTipHash();
        if (hash != null) {
            ApproveeViewModel approvers = ApproveeViewModel.load(tangle, hash);
            if (approvers.size() != 0) {
                tipsViewModel.removeTipHash(hash);
            } else if (transactionValidator.checkSolidity(hash, false)) {
                tipsViewModel.setSolid(hash);
            }
        }
    }
```

**Comment by footloosejava** [comment]
Some while later now and ...

> 05/22 19:22:22.430 [Tip Solidity Rescan] INFO  com.iota.iri.service.TipsManager - Last 100 scans - avg time= 1570 ms.

The time has gotten slower over time.

**Comment by gjeee** [comment]
i will do the same experiment....hold on :-)

**Comment by gjeee** [comment]
ok on a side-note i already know 1 optimization.....check out the function responsible for picking a random non-solid tip TipsViewModel.getRandomNonSolidTipHash()

that is very inefficient to say the least especially when this queue fills up....it has to do on avg 2500 loops just to pick a element......random access container would be more suitable.

**Comment by footloosejava** [comment]
> random access container would be more suitable.
Sure. But then it wouldn't be a LinkedHashSet.

The best you can do is iterate a set unless you want to penalize yourself every time you insert an element into a random access list - checking the equals for every element.

The alternative would be to break the set up into multiple sets with some degree of random access like a List<Set>. Then you could maintain a size function and when wanting an indexed element, then skip right to the actual Set that would contain those indexes.

But for a Set this is as good as it gets.

**Comment by gjeee** [comment]
regarding the container.....what do you thnk about this implementation? It's a bit like a hybrid:

[ListOrderedSet](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/set/ListOrderedSet.html)

**Comment by footloosejava** [comment]
The code you pasted does not compile. 'N' is out of sight. Also, you don't need to check size() - it already does that or else it returns null.

Try this and then let us know what you get after 10 to 15 minutes:

```
    private void scanTipsForSolidity() throws Exception {
        long time = System.nanoTime();
        Hash hash = tipsViewModel.getRandomNonSolidTipHash();
        if (hash != null) {
            ApproveeViewModel approvers = ApproveeViewModel.load(tangle, hash);
            if (approvers.size() != 0) {
                tipsViewModel.removeTipHash(hash);
            } else if (transactionValidator.checkSolidity(hash, false)) {
                tipsViewModel.setSolid(hash);
            }
        }
        log.info((System.nanoTime() - time) + "ns per scan");
    }
```

**Comment by footloosejava** [comment]
That set looks interesting. But my money is on the Transaction loading and backend database as being the bottleneck. I suggest timing the set random retrieve function separately to see what % of the operation it takes compared to the rest of the function.

**Comment by gjeee** [comment]
i know i know i pasted the wrong code :-(      
the code i used to produce the pasted result is:
```

    public void init() {
        solidityRescanHandle = new Thread(() -> {
            long totalTime = 0;
            long count = 0;
            while (!shuttingDown) {
                try {
                    count++;
                    long time = System.nanoTime();
                    scanTipsForSolidity();
                    totalTime += (System.nanoTime() - time);
                    if (count == 100) {
                        log.info("Last {} scans - avg time= {} ms.", count, (int) (totalTime / count));
                        count = 0;
                        totalTime = 0;
                    }
                } catch (Exception e) {
                    log.error("Error during solidity scan : {}", e);
                }
                try {
                    Thread.sleep(RESCAN_TX_TO_REQUEST_INTERVAL);
                } catch (InterruptedException e) {
                    log.error("Solidity rescan interrupted.");
                }
            }
        }, "Tip Solidity Rescan");
        solidityRescanHandle.start();
    }

    private void scanTipsForSolidity() throws Exception {
       
        long time = System.nanoTime();
        int n=0;
	for (n = 0; n < 1 && tipsViewModel.nonSolidSize() != 0; n++) {
            Hash hash = tipsViewModel.getRandomNonSolidTipHash();
            if (hash != null) {
                ApproveeViewModel approvers = ApproveeViewModel.load(tangle, hash);
                if (approvers.size() != 0) {
                    tipsViewModel.removeTipHash(hash);
                } else if (transactionValidator.checkSolidity(hash, false)) {
                    tipsViewModel.setSolid(hash);
                }
            }
        }
        if(n>0){
	    log.info((System.nanoTime() - time)/n + "ns per tx");
	}
    }
```

**Comment by footloosejava** [comment]
>Last 100 scans - avg time= 1195517 **ns**

But your logging line says:

```
log.info("Last {} scans - avg time= {} ms.", count, (int) (totalTime / count));
```

**Comment by gjeee** [comment]
true i changed that the hard way after copy/pasting :-)
you are sharp for sure :-)

**Comment by gjeee** [comment]
preliminary results show that the getRandomNonSolidTipHash() call makes up approx 4% of the total scanTipsForSolidity() call....so not the first candidate for optimization :-)

**Comment by gjeee** [comment]
@footloosejava i think we can only compare similar functioning nodes......if your node is not fully synced it misses a lot of tx and thus gaps in the branches which lead to expensive deep-scan routines.

was your node fully synced when you conducted your test?

**Comment by footloosejava** [comment]
Not fully synced.

Though ... thinking about it critically, the node should be designed to be as fast as possible at the times when it is slowest - such as when it is not fully synced. If the network only performs when it is stable ... then it won't stand up to much adversity.

Moar coffee now :)

**Comment by gjeee** [comment]
solved with PR #913
